### DISCUSSION TITLE
Is there a way to password protect my next.js app while in development

### DISCUSSION THREAD

**Discussion by jaapbakker88** [vercel/next.js#12130] [Help] [answered]
I can't find out if there's a way to put a simple password protect on the next.js app I'm building. Basically I just want the people involved in the project to see the progress I'm making right now, but have it available on the deploy domains on Vercel. What would the best way of achieving this be?

**Comment by c0b41** [comment]
custom http header with basic http auth may help you 
https://vercel.com/docs/configuration#routes/headers

**Reply by jaapbakker88** [reply]
Thanks, that's a start but I'm still not sure how to implement this so you need a password to view the ongoing development. If I set the header and http basic auth it's already authenticated when I visit the page. Tried using a custom server, which worked locally but then I found out the custom server is not supported on the Vercel deploys.

**Reply by c0b41** [reply]
@jaapbakker88 yeah, look's like this feature backloged https://github.com/zeit/now/issues/2869

**Comment by wiiiimm** [comment]
@jaapbakker88 so the team at Vercel is working on a system-level password protection right now and they will make an official announcement about it when the feature is ready. For now, you can have a look at this unofficial repo I built on the side: https://github.com/bbiHQ/simple-basic-http-auth-proxy-vercel. It takes in ORIGIN, USERNAME, PASSWORD from env and checks for credentials, if everything checkout, it proxy your requests to your stage deployment (ORIGIN).

**Reply by jaapbakker88** [reply]
Thanks @williamli

**Reply by wiiiimm** [reply]
@jaapbakker88 Vercel has enabled native password and SSO protection at system level.

https://github.com/zeit/next.js/discussions/12130#discussioncomment-8536

**Reply by wereHamster** [reply]
Note that this is only available for paid plans (150$ / month)!

**Comment by wiiiimm** [comment]
Hi @jaapbakker88, we have just enabled the password protection and SSO protection features for Pro and Enterprise plans. Please take a look at https://vercel.com/blog/protecting-deployments for more information.

**Reply by jaapbakker88** [reply]
Thanks again! @williamli

**Reply by darrylmorley** [reply]
$150 per month?! 🤣😭

**Reply by wiiiimm** [reply]
> $150 per month?! 🤣😭

yes, it's an account-wide feature so once enabled, you can use it across on all your deployments and all your projects.
the password protection happens on the edge level so it locks both your static frontend as well as serverless functions.

**Reply by JonCognioDigital** [reply]
For those of use that have a Vercel account to host a single site this price is absolutely insane. Having signed up to Vercel I now find myself in a position where I either have to move everything to another hosting provider, or pay $150 every month just to password protect a development site. Seriously not cool!

**Comment by dcantatore** [comment]
What makes this even worse is that they hide the price behind creating a Pro level account at 20/month and then after you do that and assume it will be there when you go back in they try and get the 150/month from you. 

1. That price is too high when you can just add a middle ware:
```https://gist.github.com/mtt87/cb9f36665165112468b92aa9fb470c7e```
2. Don't try to be sneaky because all you do is make devs angry for no reason

**Reply by GabenGar** [reply]
> That price is too high when you can just add a middle ware

Aren't middlewares too brittle for prod, especially for pages?

**Comment by wereHamster** [comment]
This can be done via Next.js middlewares, there is even an official example: https://github.com/vercel/examples/tree/main/edge-functions/basic-auth-password.

No more paying 150$/month for this basic auth (pun intended) feature!

**Comment by leerob** [comment, marked as answer]
Hey folks, wanted to follow up here!

- [You can now protect preview deployments for free with Vercel authentication](https://vercel.com/blog/security-controls-protected-preview-deployments-passwords)
- [You can also do basic auth with a password fairly easy with Middleware](https://github.com/vercel/examples/tree/main/edge-functions/basic-auth-password)
